File: app/src/main/java/com/drdisagree/colorblendr/utils/monet/ColorSchemeUtil.java
package com.drdisagree.colorblendr.utils.monet;

import android.content.Context;

import androidx.annotation.ColorInt;

import com.drdisagree.colorblendr.R;
import com.drdisagree.colorblendr.utils.SystemUtil;
import com.drdisagree.colorblendr.utils.monet.hct.Hct;
import com.drdisagree.colorblendr.utils.monet.palettes.TonalPalette;
import com.drdisagree.colorblendr.utils.monet.scheme.DynamicScheme;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeContent;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeExpressive;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeFidelity;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeFruitSalad;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeMonochrome;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeNeutral;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeRainbow;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeTonalSpot;
import com.drdisagree.colorblendr.utils.monet.scheme.SchemeVibrant;

import java.util.ArrayList;

public class ColorSchemeUtil {

    public enum MONET {
        NEUTRAL,
        MONOCHROME,
        TONALSPOT,
        VIBRANT,
        RAINBOW,
        EXPRESSIVE,
        FIDELITY,
        CONTENT,
        FRUIT_SALAD
    }

    public static final int[] tones = {100, 99, 95, 90, 80, 70, 60, 50, 40, 30, 20, 10, 0};

    public static ArrayList<ArrayList<Integer>> generateColorPalette(MONET style, @ColorInt int color) {
        ArrayList<ArrayList<Integer>> palette = new ArrayList<>();

        DynamicScheme dynamicScheme = switch (style) {
            case NEUTRAL -> new SchemeNeutral(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case MONOCHROME -> new SchemeMonochrome(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case TONALSPOT -> new SchemeTonalSpot(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case VIBRANT -> new SchemeVibrant(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case RAINBOW -> new SchemeRainbow(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case EXPRESSIVE -> new SchemeExpressive(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case FIDELITY -> new SchemeFidelity(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case CONTENT -> new SchemeContent(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
            case FRUIT_SALAD ->
                    new SchemeFruitSalad(Hct.fromInt(color), SystemUtil.isDarkMode(), 5);
        };

        TonalPalette[] palettes = {
                dynamicScheme.primaryPalette,
                dynamicScheme.secondaryPalette,
                dynamicScheme.tertiaryPalette,
                dynamicScheme.neutralPalette,
                dynamicScheme.neutralVariantPalette
        };

        for (TonalPalette paletteType : palettes) {
            palette.add(createToneList(paletteType));
        }

        return palette;
    }

    private static ArrayList<Integer> createToneList(TonalPalette palette) {
        ArrayList<Integer> toneList = new ArrayList<>();
        for (int tone : ColorSchemeUtil.tones) {
            toneList.add(palette.tone(tone));
        }
        return toneList;
    }

    public static MONET stringToEnum(Context context, String enumString) {
        if (enumString.equals(context.getString(R.string.monet_neutral))) {
            return MONET.NEUTRAL;
        } else if (enumString.equals(context.getString(R.string.monet_monochrome))) {
            return MONET.MONOCHROME;
        } else if (enumString.equals(context.getString(R.string.monet_tonalspot))) {
            return MONET.TONALSPOT;
        } else if (enumString.equals(context.getString(R.string.monet_vibrant))) {
            return MONET.VIBRANT;
        } else if (enumString.equals(context.getString(R.string.monet_rainbow))) {
            return MONET.RAINBOW;
        } else if (enumString.equals(context.getString(R.string.monet_expressive))) {
            return MONET.EXPRESSIVE;
        } else if (enumString.equals(context.getString(R.string.monet_fidelity))) {
            return MONET.FIDELITY;
        } else if (enumString.equals(context.getString(R.string.monet_content))) {
            return MONET.CONTENT;
        } else if (enumString.equals(context.getString(R.string.monet_fruitsalad))) {
            return MONET.FRUIT_SALAD;
        } else {
            return MONET.TONALSPOT;
        }
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/utils/colors/ColorSchemeUtil.kt
package com.drdisagree.colorblendr.utils.colors

import android.content.Context
import androidx.annotation.ColorInt
import com.drdisagree.colorblendr.ColorBlendr.Companion.appContext
import com.drdisagree.colorblendr.R
import com.drdisagree.colorblendr.data.common.Utilities
import com.drdisagree.colorblendr.data.enums.MONET
import com.drdisagree.colorblendr.utils.app.MiscUtil.getOriginalString
import com.drdisagree.colorblendr.utils.app.SystemUtil
import com.drdisagree.materialcolorutilities.dynamiccolor.ColorSpec
import com.drdisagree.materialcolorutilities.dynamiccolor.DynamicScheme
import com.drdisagree.materialcolorutilities.hct.Hct
import com.drdisagree.materialcolorutilities.palettes.TonalPalette
import com.drdisagree.materialcolorutilities.scheme.SchemeCmf
import com.drdisagree.materialcolorutilities.scheme.SchemeContent
import com.drdisagree.materialcolorutilities.scheme.SchemeExpressive
import com.drdisagree.materialcolorutilities.scheme.SchemeFidelity
import com.drdisagree.materialcolorutilities.scheme.SchemeFruitSalad
import com.drdisagree.materialcolorutilities.scheme.SchemeMonochrome
import com.drdisagree.materialcolorutilities.scheme.SchemeNeutral
import com.drdisagree.materialcolorutilities.scheme.SchemeRainbow
import com.drdisagree.materialcolorutilities.scheme.SchemeTonalSpot
import com.drdisagree.materialcolorutilities.scheme.SchemeVibrant

object ColorSchemeUtil {

    private val tones: IntArray = intArrayOf(100, 99, 95, 90, 80, 70, 60, 50, 40, 30, 20, 10, 0)

    fun generateColorPalette(
        style: MONET,
        @ColorInt color: Int,
        isDark: Boolean = SystemUtil.isDarkMode,
        contrast: Int = 0
    ): ArrayList<ArrayList<Int>> {
        val palette = ArrayList<ArrayList<Int>>()
        val specVersion: ColorSpec.SpecVersion = when (Utilities.getColorSpecVersion()) {
            0 -> ColorSpec.SpecVersion.SPEC_2021
            1 -> ColorSpec.SpecVersion.SPEC_2025
            2 -> ColorSpec.SpecVersion.SPEC_2026
            else -> ColorSpec.SpecVersion.SPEC_2021
        }
        val platform: DynamicScheme.Platform = DynamicScheme.DEFAULT_PLATFORM

        val dynamicScheme = getDynamicScheme(
            style,
            color,
            isDark,
            contrast,
            specVersion,
            platform
        )

        val tonalPalettes = arrayOf(
            dynamicScheme.primaryPalette,
            dynamicScheme.secondaryPalette,
            dynamicScheme.tertiaryPalette,
            dynamicScheme.neutralPalette,
            dynamicScheme.neutralVariantPalette,
            dynamicScheme.errorPalette
        )

        for (tonalPalette in tonalPalettes) {
            palette.add(createToneList(tonalPalette))
        }

        return palette
    }

    private fun getDynamicScheme(
        style: MONET,
        @ColorInt color: Int,
        isDark: Boolean,
        contrast: Int,
        specVersion: ColorSpec.SpecVersion,
        platform: DynamicScheme.Platform
    ): DynamicScheme {
        return when (style) {
            MONET.SPRITZ -> SchemeNeutral(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.MONOCHROMATIC -> SchemeMonochrome(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.TONAL_SPOT -> SchemeTonalSpot(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.VIBRANT -> SchemeVibrant(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.RAINBOW -> SchemeRainbow(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.EXPRESSIVE -> SchemeExpressive(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.FIDELITY -> SchemeFidelity(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.CONTENT -> SchemeContent(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.FRUIT_SALAD -> SchemeFruitSalad(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )

            MONET.CMF -> SchemeCmf(
                sourceColorHct = Hct.fromInt(color),
                isDark = isDark,
                contrastLevel = contrast.toDouble(),
                specVersion = specVersion,
                platform = platform
            )
        }
    }

    private fun createToneList(palette: TonalPalette): ArrayList<Int> {
        val toneList = ArrayList<Int>()
        for (tone in tones) {
            toneList.add(palette.tone(tone))
        }
        return toneList
    }

    fun Int.getStyleNameForRootless(): String {
        return when (this.getOriginalString()) {
            R.string.monet_neutral.getOriginalString() -> "SPRITZ"
            R.string.monet_vibrant.getOriginalString() -> "VIBRANT"
            R.string.monet_expressive.getOriginalString() -> "EXPRESSIVE"
            R.string.monet_rainbow.getOriginalString() -> "RAINBOW"
            R.string.monet_fruitsalad.getOriginalString() -> "FRUIT_SALAD"
            R.string.monet_content.getOriginalString() -> "CONTENT"
            R.string.monet_monochrome.getOriginalString() -> "MONOCHROMATIC"
            R.string.monet_fidelity.getOriginalString() -> "FIDELITY"
            else -> "TONAL_SPOT"
        }
    }

    fun stringToEnumMonetStyle(
        context: Context = appContext,
        enumString: String
    ): MONET {
        // compare both original string and localized string
        return when (enumString) {
            R.string.monet_neutral.getOriginalString(),
            context.getString(R.string.monet_neutral) -> MONET.SPRITZ

            R.string.monet_monochrome.getOriginalString(),
            context.getString(R.string.monet_monochrome) -> MONET.MONOCHROMATIC

            R.string.monet_tonalspot.getOriginalString(),
            context.getString(R.string.monet_tonalspot) -> MONET.TONAL_SPOT

            R.string.monet_vibrant.getOriginalString(),
            context.getString(R.string.monet_vibrant) -> MONET.VIBRANT

            R.string.monet_rainbow.getOriginalString(),
            context.getString(R.string.monet_rainbow) -> MONET.RAINBOW

            R.string.monet_expressive.getOriginalString(),
            context.getString(R.string.monet_expressive) -> MONET.EXPRESSIVE

            R.string.monet_fidelity.getOriginalString(),
            context.getString(R.string.monet_fidelity) -> MONET.FIDELITY

            R.string.monet_content.getOriginalString(),
            context.getString(R.string.monet_content) -> MONET.CONTENT

            R.string.monet_fruitsalad.getOriginalString(),
            context.getString(R.string.monet_fruitsalad) -> MONET.FRUIT_SALAD

            else -> MONET.TONAL_SPOT
        }
    }
}